### PR TITLE
Added models and their validations

### PR DESCRIPTION
**Pull request by Cena-JM**

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -3,4 +3,20 @@ class User < ApplicationRecord
...
+  has_many :likes
+  has_many :comments
+
+  has_many :friend_requests, dependent: :destroy
+  has_many :pending_friends, through: :friend_requests, source: :requested
+
+  has_many :friendships, dependent: :destroy
+  has_many :friends, through: :friendships, class_name: 'User'
+
+  validates :name, presence: true, length: { maximum: 50 }
```
Comment (rvvergara) [inline app/models/user.rb]: I tried to create a user with the attributes as you specified in your validations:

image

But as you see I got an error. It appears you need to work on your user model still. 

It appears you are putting an attribute in your user model that you didn't specify in your schema. Please fix this.

File: db/schema.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+#
+# It's strongly recommended that you check this file into your version control system.
+
+ActiveRecord::Schema.define(version: 2019_05_09_094314) do
+
+  # These are extensions that must be enabled in order to support this database
+  enable_extension "plpgsql"
+
+  create_table "users", force: :cascade do |t|
+    t.string "email", default: "", null: false
```
Comment (rvvergara) [inline db/schema.rb]: I don't see the column `name` in here but you are using it in the User model.

File: app/models/friendship.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+class Friendship < ApplicationRecord
```
Comment (rvvergara) [inline app/models/friendship.rb]: I tried to do the ff:

Created 2 users -> `user1` and `user2`.. Then I created the following friendship:

`user1.friendships.create(friend_id: user2.id)`

So I would expect now `user1.friends` to include `user2` right? And so I ran:
`user1.friends` -> And yes `user2` was there.

But then I also would expect that running `user2.friends` I would see `user1` right?

But that's not the case:
`user2.friends` results in an empty array

File: test/models/friendship_test.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+require 'test_helper'
```
Comment (rvvergara) [inline test/models/friendship_test.rb]: Please also include here test for association with the User model.

You can actually use the shoulda-matchers gem to make this association test easier. https://guides.railsgirls.com/testing-shoulda-matchers

File: test/models/user_test.rb [hunk 5]
```diff
@@ -1,7 +1,68 @@
 require 'test_helper'
```
Comment (rvvergara) [inline test/models/user_test.rb]: Pls add association tests for the user. Please use shoulda-gem as I shared in the friendship test comment

File: test/models/friend_request_test.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
+require 'test_helper'
```
Comment (rvvergara) [inline test/models/friend_request_test.rb]: Please add test for associations here.. You may use shoulda-gem I mentioned in my comment on friendship_test

File: app/models/friendship.rb (hunk 3 above)
Comment (kvnlay) [reply app/models/friendship.rb]: Screen Shot 2019-05-14 at 11 32 50 AM
I and my partner tried testing friendships on our console and it works fine.
1. I created users `u1` and `u2`
2. I created the friendship `u1.friendships.create(friend_id: u2.id)`
3. I checked for u1 friends by running `u1.friends` and I got an array with `u2` in it
4. I also ran `u2.friends` and I got an array with `u1` in it

File: app/models/user.rb [hunk 7]
```diff
@@ -1,6 +1,27 @@
...
+  before_save { email.downcase! }
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
+  has_many :posts
+  has_many :likes
+  has_many :comments
+
+  has_many :recieved_requests, foreign_key: "requested_id",
```
Comment (rvvergara) [inline app/models/user.rb]: Haha I wanna ask you guys a favor, can we correct the spelling of `recieved`? It's supposed to be `received`.

I have a suggestion too, but implementing this is entirely up to you and depending of course on your implementation of friend requests later on in the controller and views. For friendships I can do something like `user1.friends` right? How about if I wanna know how many people `user1` has sent a request to? Or how many requesters `user1` has? I mean right now obviously we can use `sent_requests` and `received_requests`, but for these two methods we still need to type either `requester_id` or `requested_id` in a conditional to get what we want. Probably a method or an additional `has_many` association for requested and requester is possible? Similar to the friendships and friends association you already have.

File: test/models/post_test.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
+require 'test_helper'
```
Comment (rvvergara) [inline test/models/post_test.rb]: Let's add a test for when a post is deleted comments and likes of that post should also be deleted.

File: test/models/user_test.rb [hunk 9]
```diff
@@ -1,7 +1,81 @@
...
 
 class UserTest < ActiveSupport::TestCase
-  # test "the truth" do
-  #   assert true
-  # end
+  def setup
```
Comment (rvvergara) [inline test/models/user_test.rb]: Let's just add tests for when a user is deleted, the associated friendships, requests, posts, comments and likes should also be deleted

File: app/models/comment.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
+
+  validates :content, presence: true, length: { maximum: 60 }
+  validates :user_id, presence: true
```
Comment (rvvergara) [inline app/models/comment.rb]: We already have the association established in lines 2-3. We don't need anymore to validate presence of user_id and post_id. :)

File: app/models/post.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+class Post < ApplicationRecord
+  belongs_to :user
+  has_many :comments
+  has_many :likes
+
+  validates :user_id, presence: true
```
Comment (rvvergara) [inline app/models/post.rb]: We already have the association with user established in line 2, we don't need anymore to validate presence of user_id here as no post will be saved unless it's created with an existing user

File: app/models/friend_request.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+class FriendRequest < ApplicationRecord
+  belongs_to :requester, class_name: 'User'
+  belongs_to :requested, class_name: 'User'
+
+  validates :requester, presence: true
```
Comment (rvvergara) [inline app/models/friend_request.rb]: We don't need to validate presence of requester and requested. The associations in lines 2-3 themselves already validate the presence of these two attributes.

File: test/models/friend_request_test.rb [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+class FriendRequestTest < ActiveSupport::TestCase
+  def setup
+    @requester = users(:musa)
+    @requested = users(:kingsley)
+    @friend_request = FriendRequest.new(requester_id: @requester.id,\
+                                        requested_id: @requested.id)
+  end
+
+  test 'should be valid' do
```
Comment (rvvergara) [inline test/models/friend_request_test.rb]: We can safely remove these validation tests as we can also remove their corresponding code in the model (see my comments in the request model). I realized this just now that the associations themselves act as validators. Therefore, your association tests using shoulda already took care of these validations

File: test/models/friendship_test.rb [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+class FriendshipTest < ActiveSupport::TestCase
+  def setup
+    @user = users(:musa)
+    @friend = users(:kingsley)
+    @friendship = Friendship.new(user_id: @user.id, friend_id: @friend.id)
+  end
+
+  test 'should be valid' do
```
Comment (rvvergara) [inline test/models/friendship_test.rb]: We can safely remove the tests for the presence of friend_id, user_id as the association tests themselves take care of validating the presence of these two attributes in a friendship row.

File: test/models/post_test.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+  def setup
+    @user = users(:musa)
+    @post = Post.new(body: 'Lorem ipsum', user_id: @user.id)
+  end
+
+  test 'should be valid' do
+    assert @post.valid?
+  end
+
+  test 'user id should be present' do
```
Comment (rvvergara) [inline test/models/post_test.rb]: We can safely remove testing for this as the association test below handles the presence of user_id

File: test/models/comment_test.rb [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    @post = posts(:sample_post1)
+    @comment = Comment.new(content: 'Lorem ipsum', user_id: @user.id,\
+                           post_id: @post.id)
+  end
+
+  test 'should be valid' do
+    assert @comment.valid?
+  end
+
+  test 'user id should be present' do
```
Comment (rvvergara) [inline test/models/comment_test.rb]: We can safely remove testing for the presence of user_id as the association test below takes care of this already

File: test/models/comment_test.rb [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    @comment.user_id = nil
+    assert_not @comment.valid?
+  end
+
+  test 'post id should be present' do
```
Comment (rvvergara) [inline test/models/comment_test.rb]: We can safely remove testing for the presence of post_id as the association test below takes care of this already

File: app/models/like.rb [hunk 18]
```diff
@@ -0,0 +1,7 @@
+class Like < ApplicationRecord
+  validates :user, presence: true
```
Comment (rvvergara) [inline app/models/like.rb]: We don't need to validate the presence of user and post anymore as our belongs_to associations take care of that for us

File: test/models/like_test.rb [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+    @user = users(:musa)
+    @post = posts(:sample_post1)
+    @like = Like.new(user_id: @user.id, post_id: @post.id)
+  end
+
+  test 'should be valid' do
+    assert @like.valid?
+  end
+
+  test 'user id should be present' do
```
Comment (rvvergara) [inline test/models/like_test.rb]: We can safely delete this test case. The association test takes care of this

File: test/models/like_test.rb [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+    @like.user_id = nil
+    assert_not @like.valid?
+  end
+
+  test 'post id should be present' do
```
Comment (rvvergara) [inline test/models/like_test.rb]: We can safely delete this test case. The association test takes care of this

File: app/models/user.rb [hunk 21]
```diff
@@ -1,6 +1,29 @@
...
+  has_many :likes, dependent: :destroy
+  has_many :comments, dependent: :destroy
+
+  has_many :received_requests, foreign_key: "requested_id", 
+                                class_name: "FriendRequest", 
+                                dependent: :destroy
+  has_many :sent_requests, foreign_key: "requester_id", 
+                            class_name: "FriendRequest", 
+                            dependent: :destroy
+  has_many :requesters, through: :received_requests, source: :requester
```
Comment (rvvergara) [inline app/models/user.rb]: Great!!

File: app/models/user.rb [hunk 22]
```diff
@@ -1,6 +1,29 @@
...
+  before_save { email.downcase! }
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
+  has_many :posts, dependent: :destroy
+  has_many :likes, dependent: :destroy
+  has_many :comments, dependent: :destroy
+
+  has_many :received_requests, foreign_key: "requested_id",
```
Comment (rvvergara) [inline app/models/user.rb]: Thanks for correcting the previous typo

File: test/models/post_test.rb [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+    assert_not @post.valid?
+  end
+
+  test 'body should be at most 140 characters' do
+    @post.body = 'a' * 201
+    assert_not @post.valid?
+  end
+
+  should belong_to(:user)
+  should have_many(:comments)
```
Comment (rvvergara) [inline test/models/post_test.rb]: Great job including the dependent destroy here

**Pull request merged**